### PR TITLE
Add a speed modifier slider

### PR DESCRIPTION
**Pull request by Plextora**
https://github.com/user-attachments/assets/0f391949-1053-4469-aea2-3b4907ec24c2

Kinda self explanatory, this PR adds an extra slider next to the Volume slider that change's the speed of the music. This nice for songs that are famous DT maps, like Tsukinami or rog-unlimitation.

### CODE REVIEW

**Review by hrfarmer — CHANGES_REQUESTED** [review]
delete the package-lock, we moved to using bun for managing packages in #97

**Comment by duduBTW** [comment]
@Plextora 

Can we replace the slider by the volume with a popover on the right?


https://github.com/user-attachments/assets/755a4741-1da1-4290-bf78-70fbf1c8dfb0

### Code

### Popover

Add a popover trigger.

`src\renderer\src\components\popover\PopoverTrigger.tsx`
``` tsx
type PopoverAnchorProps = JSX.IntrinsicElements["div"];
export const PopoverAnchor: Component<PopoverAnchorProps> = (props) => {
  const state = usePopover();
  return <div ref={state.setTriggerRef} {...props} />;
};
```

Add the trigger to the popover component.

`src\renderer\src\components\popover\Popover.tsx`
``` tsx
const Popover = Object.assign(PopoverRoot, {
  Anchor: PopoverAnchor,
  Content: PopoverContent,
  Trigger: PopoverTrigger,
  Overlay: PopoverOverlay,
});
```

on line 42:
``` diff
- const [triggerRef, _setTriggerRef] = createSignal<HTMLButtonElement | null>(null);
+ const [triggerRef, _setTriggerRef] = createSignal<HTMLElement | null>(null);
```

---

### Song Controls

Move the button to the right part and implement the popover.

`src\renderer\src\components\song\song-detail\SongControls.tsx`

Remove the component from the `LeftPart`.

Add it to the `RightPart`.

``` tsx

const PREDEFINED_SPEEDS: number[] = [0.25, 0.5, 1, 1.5, 2] as const;
const MIN_SPEED_AMOUNT = PREDEFINED_SPEEDS[0];
const MAX_SPEED_AMOUNT = PREDEFINED_SPEEDS.at(-1);
const RightPart = () => {
  const [isPopoverOpen, setisPopoverOpen] = createSignal(false);

  return (
    <div class="flex flex-1 justify-end gap-4">
      <Popover
        isOpen={isPopoverOpen}
        onValueChange={setisPopoverOpen}
        placement="top-end"
        offset={{
          mainAxis: 10,
        }}
      >
        <Popover.Anchor>
          <Button
            onClick={() => setisPopoverOpen(true)}
            size="icon"
            variant="ghost"
            title="Set speed"
          >
            <GaugeIcon size={20} />
          </Button>
        </Popover.Anchor>

        <Portal>
          <Popover.Overlay />
          <Popover.Content class="flex w-fit min-w-48 flex-col  rounded-xl ring-stroke ring-1 ring-inset bg-thick-material px-1.5 py-3 backdrop-blur-md shadow-xl">
            <p class="font-medium text-sm text-subtext gap-1 px-2">Custom Speed</p>
            <div class="flex flex-col px-2">
              <Slider
                class="flex h-8 flex-grow items-center"
                min={MIN_SPEED_AMOUNT}
                max={MAX_SPEED_AMOUNT}
                value={speed}
                onValueChange={setSpeed}
                enableWheelSlide
              >
                <Slider.Track class="h-1 flex-1 rounded bg-thick-material ring-1 ring-stroke">
                  <Slider.Range class="block h-1 rounded bg-white" />
                </Slider.Track>
                <Slider.Thumb class="mt-2 block h-4 w-4 rounded-full bg-white" />
              </Slider>
              <div class="text-xs ">{Math.round(speed() * 100) / 100}x</div>
            </div>
            <div class="w-full bg-stroke h-px my-2" />

            <For each={PREDEFINED_SPEEDS}>
              {(amount) => <SpeedOption amount={amount}>{amount}</SpeedOption>}
            </For>
          </Popover.Content>
        </Portal>
      </Popover>
      <Button size="icon" variant="ghost">
        <CirclePlusIcon size={20} />
      </Button>
    </div>
  );
};

type SpeedOptionProps = {
  amount: number;
};
const SpeedOption: ParentComponent<SpeedOptionProps> = (props) => {
  return (
    <button
      onClick={() => setSpeed(props.amount)}
      class="flex items-center justify-between rounded-md px-2 py-1.5 disabled:opacity-50 disabled:pointer-events-none focus:outline-none hover:bg-surface"
    >
      {props.children}
    </button>
  );
};
```

**Review by duduBTW — DISMISSED** [review]
Requested design changes.

File: src/renderer/src/components/song/song.utils.ts [hunk 1]
```diff
@@ -295,6 +303,7 @@ window.api.listen("queue::songChanged", async (s) => {
   setSong(s);
   await window.api.request("discord::play", s);
   await play();
+  player.playbackRate = speedNum;
```
Comment (duduBTW) [inline src/renderer/src/components/song/song.utils.ts:306]: Why do we have two variables to track the speed, can't we just use the "speed" signal here?

Comment (Plextora) [reply src/renderer/src/components/song/song.utils.ts]: Thanks for catching that, I accidentally kept that in there while debugging something else. This should be fixed in https://github.com/Team-BTMC/osu-radio/pull/103/commits/d75d755d27420d35a91a567ceb08539efc5085b9

**Review by hrfarmer — APPROVED** [review]

**Pull request merged**